apps/drupal/src/Plugin/Preprocessor/PreprocessorManager.php

```diff
@@ -0,0 +1,500 @@
+<?php
+
+namespace Drupal\particle\Plugin\Preprocessor;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\particle\Tools\ParticleTools;
+
+/**
+ * Provides a theme preprocessor manager for Class based theme preprocessors.
+ *
+ * Drupal core does not provide built-in plugin support for theme files.
+ * The plugin discovery might before the front end theme is not active.
+ *
+ * The approach here is to use Drupal's State API to handle Preprocessor Class.
+ * New suggestions are read on page load and added to State API via STATE_ID.
+ * The Preprocessor Class Map is stored in cache and refreshed on change.
+ *
+ */
+class PreprocessorManager {
+
+  /**
+   * The State API ID for this Manager.
+   *
+   * @var string
+   */
+  const STATE_ID = 'preprocessor.map';
+
+  /**
+   * Appended suggestions.
+   *
+   * @see getSuggestions()
+   *
+   * @var array
+   */
+  protected static $allSuggestions = [];
+
+  /**
+   * Cached Preprocessor instances.
+   *
+   * @see load()
+   *
+   * @var array
+   */
+  protected static $preprocessorInstances = [];
+
+  /**
+   * Particle tools instance for all preprocessors.
+   *
+   * @var Drupal\particle\Tools\ParticleTools
+   *   Particle's suite of helper functions and utilities.
+   */
+  protected static $tools;
+
+  /**
+   * Returns the tools object.
+   *
+   * @return Drupal\particle\Tools\ParticleTools
+   *   The tools object.
+   */
+  public static function getTools() {
+    if (!isset(static::$tools)) {
+      static::$tools = new ParticleTools();
+    }
+
+    return static::$tools;
+  }
+
+  /**
+   * Returns the full path to the directory of the preprocessor classes.
+   *
+   * Base should be the only root level preprocessor in this directory.
+   * All other types should be located in subdirectories.
+   *
+   * @see getPreprocessors()
+   *
+   * @return string
+   *   The directory path.
+   */
+  public static function getPreprocessorsDirectory() {
+    return __DIR__ . '/Preprocessors';
+  }
+
+  /**
+   * Get the preprocessors from directory and subdirectories.
+   *
+   * @return array
+   *   An array of preprocessors from the file system.
+   */
+  public static function getPreprocessors() {
+    $preprocessors = [];
```

Comment:
@jryanconklin Code suggestion: Cache $preprocessors with a static class property or drupal_static() so that all directories are not scanned each time new suggestions are found for mergedMaps().

Reply:
Updated! Thanks for the suggestion!